internal/pkg/agent/application/enroll/enroll.go

```diff
@@ -0,0 +1,574 @@
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package enroll
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"math/rand/v2"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/otiai10/copy"
+	"gopkg.in/yaml.v2"
+
+	"github.com/elastic/elastic-agent-libs/file"
+	"github.com/elastic/elastic-agent-libs/transport/tlscommon"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/filelock"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/info"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/paths"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/configuration"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/errors"
+	"github.com/elastic/elastic-agent/internal/pkg/config"
+	"github.com/elastic/elastic-agent/internal/pkg/core/backoff"
+	monitoringConfig "github.com/elastic/elastic-agent/internal/pkg/core/monitoring/config"
+	"github.com/elastic/elastic-agent/internal/pkg/crypto"
+	"github.com/elastic/elastic-agent/internal/pkg/fleetapi"
+	fleetclient "github.com/elastic/elastic-agent/internal/pkg/fleetapi/client"
+	"github.com/elastic/elastic-agent/internal/pkg/release"
+	"github.com/elastic/elastic-agent/internal/pkg/remote"
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+)
+
+const (
+	EnrollBackoffInit = time.Second * 5
+	EnrollBackoffMax  = time.Minute * 10
+
+	maxRetriesstoreAgentInfo       = 5
+	defaultFleetServerHost         = "0.0.0.0"
+	defaultFleetServerPort         = 8220
+	defaultFleetServerInternalHost = "localhost"
+	defaultFleetServerInternalPort = 8221
+	statusPath                     = "/api/status"
+	apiStatusTimeout               = 15 * time.Second
+	backupSuffix                   = ".enroll.bak"
+)
+
+type saver interface {
+	Save(io.Reader) error
+}
+
+func CheckRemote(ctx context.Context, c fleetclient.Sender) error {
+	ctx, cancel := context.WithTimeout(ctx, apiStatusTimeout)
+	defer cancel()
+
+	resp, err := c.Send(ctx, http.MethodGet, "/api/status", nil, nil, nil)
+	if err != nil {
+		return fmt.Errorf("fail to communicate with Fleet Server API client hosts: %w", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("fleet server ping returned a bad status code: %d", resp.StatusCode)
+	}
+
+	// discard body for proper cancellation and connection reuse
+	_, _ = io.Copy(io.Discard, resp.Body)
+	resp.Body.Close()
+
+	return nil
+}
+
+// BackupConfig creates a backup of currently used fleet config
+func BackupConfig() error {
+	configFile := paths.AgentConfigFile()
+	backup := configFile + backupSuffix
+
+	err := copy.Copy(configFile, backup, copy.Options{
+		PermissionControl: copy.AddPermission(0600),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to backup config file %s -> %s: %w", configFile, backup, err)
+	}
+
+	return nil
+}
+
+// RestoreConfig restores from backup if needed and signals restore was performed
```

Comment:
Is there any reason for this function to be in the `enroll` package  AFAICT, it's  called from the coordinator and cmd/run. Perhaps it could live in the `coordinator` package?